### ISSUE TITLE
FIXME: Ignore mtime (use only hash of content + permissions) while invalidating cache for ADD command

### ISSUE THREAD

**Issue by tonnzor** [docker/docker#9391]
Normally Docker uses caching for ADD command.

But this cache gets invalidated once modification date of the file changed.

It should not be the case -- only permissions and content hash (e.g. MD5 or SHA) should be used for this. It is very difficult to preserve mtime in build systems.

Live example:
```bash
cat > Dockerfile <<EOF 
FROM debian:jessie
ADD myfile /tmp/
RUN echo 'Done'.
EOF
echo "Some content" > myfile
docker build .
```

```
Sending build context to Docker daemon 3.584 kB
Sending build context to Docker daemon 
Step 0 : FROM debian:jessie
 ---> aaabd2b41e22
Step 1 : ADD myfile /tmp/
 ---> f9a488e10e68
Removing intermediate container 3e9f4ee1bc73
Step 2 : RUN echo 'Done'.
 ---> Running in b848d6d8628d
Done.
 ---> d8ba43fbbb65
Removing intermediate container b848d6d8628d
Successfully built d8ba43fbbb65
```
Check caching is there:
```bash
docker build .
```
```
Sending build context to Docker daemon 3.584 kB
Sending build context to Docker daemon 
Step 0 : FROM debian:jessie
 ---> aaabd2b41e22
Step 1 : ADD myfile /tmp/
 ---> Using cache
 ---> f9a488e10e68
Step 2 : RUN echo 'Done'.
 ---> Using cache
 ---> d8ba43fbbb65
Successfully built d8ba43fbbb65
```

Change mtime and build again:
```bash
touch myfile
docker build .
```
```
Sending build context to Docker daemon 3.584 kB
Sending build context to Docker daemon 
Step 0 : FROM debian:jessie
 ---> aaabd2b41e22
Step 1 : ADD myfile /tmp/
 ---> 8dffbae0fcab
Removing intermediate container 3697ac2594a1
Step 2 : RUN echo 'Done'.
 ---> Running in 0c32cec76e34
Done.
 ---> 5e97b52f28d7
Removing intermediate container 0c32cec76e34
Successfully built 5e97b52f28d7
```
As you see -- cache is invalidated when mtime changed BUT content stayed the same.

**Comment by kanzure** [comment]
To work around this, I have been using `touch myfile` recursively in my git repos. I also posted my solution over here: http://stackoverflow.com/questions/26531108/docker-add-cache-when-git-checkout-same-file/26612694#26612694

I agree that running `touch myfile` in git hooks all the time is a little strange.... but not the most awful thing ever.

**Comment by jakajancar** [comment]
Tonnzor says it well:

> it makes very easy to invalidate but too hard to preserve cache

Code does not usually behave differently based on atime/mtime of the files, and if it does, then caching breaks it anways already: when RUN commands are executed, they would be set to current time, but if cache is used, the time is from the previous build.

**Comment by cscetbon** [comment]
@cpuguy83 +1 on your flag proposition !

**Comment by sorpaas** [comment]
+1 for this. It's especially annoying when building docker images in a pre-receive hook of git.

**Comment by jfrazelle** [comment]
closing as dup of https://github.com/docker/docker/issues/4351

**Comment by lanoxx** [comment]
In case anyone stumbles over this issue I am leaving this note so you don't have to search through the different duplicates to find the answer:

This was **fixed in Docker 1.8** by the following pull request https://github.com/docker/docker/pull/12031
